src/_pytest/terminal.py

```diff
@@ -803,31 +806,36 @@ def getreports(self, name):
         return values
 
     def summary_warnings(self):
-        if self.hasopt("w"):
-            all_warnings = self.stats.get(
-                "warnings"
-            )  # type: Optional[List[WarningReport]]
-            if not all_warnings:
-                return
+        if not self.hasopt("w"):
+            return
 
-            final = hasattr(self, "_already_displayed_warnings")
-            if final:
-                warning_reports = all_warnings[self._already_displayed_warnings :]
-            else:
-                warning_reports = all_warnings
-            self._already_displayed_warnings = len(warning_reports)
-            if not warning_reports:
-                return
+        all_warnings = self.stats.get("warnings")  # type: Optional[List[WarningReport]]
+        if not all_warnings:
+            return
 
-            reports_grouped_by_message = (
-                collections.OrderedDict()
-            )  # type: collections.OrderedDict[str, List[WarningReport]]
-            for wr in warning_reports:
-                reports_grouped_by_message.setdefault(wr.message, []).append(wr)
+        final = hasattr(self, "_already_displayed_warnings")
+        if final:
+            warning_reports = all_warnings[self._already_displayed_warnings :]
+        else:
+            warning_reports = all_warnings
+        self._already_displayed_warnings = len(warning_reports)
+        if not warning_reports:
+            return
 
-            title = "warnings summary (final)" if final else "warnings summary"
+        grouped = (
+            collections.OrderedDict()
+        )  # type: collections.OrderedDict[str, List[WarningReport]]
```

Comment:
Should this be:

```suggestion
        )  # type: collections.OrderedDict[str, collections.OrderedDict[str, List[WarningReport]]]
```

Or is it OK to only type-hint the values?

Reply:
That is wrong here actually - fixed it in a followup in my fork already (so this needs an update).